Fix #1366: Display top unresolved answers and new feedback for dashboard explorations in a dropdown

526avijitgupta opened the pull request:
Tasklist:
- [x] Fetch top 3 unresolved answers for every exploration in user's dashboard (along with tests).
- [x] Fetch latest 3 new feedback on every exploration (along with tests).
- [x] ~~Fetch latest 3 updates (commits made by other users) on every exploration (along with tests).~~
- [x] Frontend working on web.
- [x] Frontend working on mobile.
- [x] Fix minor bugs with sorting explorations, discovered while playing around with the dashboard manually.
- [x] Figure out and fix why a few backend tests in dashboard_test.py are failing due to weird errors!
- [x] For fetching exploration dropdown stats, create separate route on backend and implement caching on frontend.

526avijitgupta commented:
@wxyxinyu Please take a look at the general approach.

wxyxinyu commented on core/domain/stats_services.py:
```diff
@@ -47,36 +48,65 @@ def get_exps_unresolved_answers_count_for_default_rule(exp_ids):
+        {
+          'exp_id_1': {
+            'count': (number of unresolved answers for this exploration),
+            'values': [
```
Could you name this something more descriptive? Maybe 'unresolved answers'.

wxyxinyu replied:
And mention in the docstring that the unresolved answers are sorted by count.

526avijitgupta replied:
Resolved both.

wxyxinyu commented on core/domain/stats_services.py:
```diff
@@ -47,36 +48,65 @@ def get_exps_unresolved_answers_count_for_default_rule(exp_ids):
+            sorted(
+                exps_answers_mapping[exp_id]['values'],
+                key=lambda a: a['count'],
+                reverse=True)[:feconf.TOP_UNRESOLVED_ANSWERS_COUNT_DASHBOARD])
```
I think this function should do the truncation of the length of the unresolved answers, since this function is independent of the dashboard. The dashboard controller should do the truncation.

With that, the constant should also be a constant within dashboard.py and not feconf.py.

526avijitgupta replied:
Agreed about (1).
For (2), any special reasons for keeping this constant in dashboard.py instead of feconf.py ?

wxyxinyu commented on core/domain/stats_services.py:
```diff
@@ -47,36 +48,65 @@ def get_exps_unresolved_answers_count_for_default_rule(exp_ids):
-        for exp_domain_object in explorations.values()
-        for state_key in exp_domain_object.states
-    ]
+    def _get_explorations_state_tuples_by_ids(exp_ids):
```
What's the purpose of pulling this out into a function? It seems like it is only used once.

526avijitgupta replied:
Yeah, is used just once. That's why took it out as a function within this function itself.
Makes the code more readable IMO. Will be happy to revert it back, if you think otherwise.

wxyxinyu commented:
The general approach looks good! Also, this function needs tests.

526avijitgupta commented:
Renamed the function. Will be working on testing this today.

wxyxinyu commented on core/domain/exp_services.py:
```diff
@@ -1139,6 +1139,31 @@ def get_next_page_of_all_non_private_commits(
     ) for entry in results], new_urlsafe_start_cursor, more)
 
 
+def get_last_updates_for_exp_ids(user_id, exp_ids):
```
Could you add a docstring for this?

526avijitgupta replied:
Done

wxyxinyu commented on core/domain/feedback_services.py:
```diff
@@ -237,6 +237,25 @@ def _get_thread_from_model(thread_model):
         thread_model.created_on, thread_model.last_updated)
 
 
+def get_feedback_messages_multi_for_exp(exploration_ids, limit=None):
```
docstring

526avijitgupta replied:
Done

wxyxinyu commented on core/domain/feedback_services.py:
```diff
@@ -237,6 +237,25 @@ def _get_thread_from_model(thread_model):
+    for exp_id in exploration_ids:
+        result[exp_id] = []
+    for message_models_exp in messages_models_exps:
+        if limit and len(result[message_models_exp.exploration_id]) < limit:
```
This could be better structured as
`if not limit or len(... < limit):`
and then you don't need the `elif not limit` part.

526avijitgupta replied:
Done

wxyxinyu commented:
I took a quick pass at the backend part of this, looks good so far.

I'll look in more detail tomorrow.

AllanYangZhou commented on core/templates/dev/head/dashboard/Dashboard.js:
```diff
@@ -16,20 +16,31 @@
       smText: 'Publish the exploration to receive statistics.'
     };
 
+    $scope.isDropdownOpen = {};
```
The name of this would suggest it's a boolean, I suggest you rename it to `isDropdownOpenMap` or something.

seanlip replied:
I agree with @AllanYangZhou. But it seems to me that only one dropdown can be open at a time, is that right? If so, why do you need an object for this? Could you just use a string variable $scope.activeExplorationId or similar (which is null if no dropdown is open)?

526avijitgupta replied:
Updated

AllanYangZhou commented on core/templates/dev/head/dashboard/Dashboard.js:
```diff
@@ -50,25 +61,55 @@ oppia.controller('Dashboard', [
       $scope.myExplorationsView = viewType;
     };
 
+    $scope.isMobileView = function() {
```
I think `checkMobileView` makes it clearer this is a function.

seanlip replied:
Just a note, we do have functions elsewhere starting with "is". If we wanted to have a variable then we would do something like $scope.viewIsMobile.

On the other hand, I'm seeing functions called isMobileView() and checkForMobileView(). That's a bit confusing, particularly the second one which seems to have side-effects. Maybe the latter should be something like updatePublishTextGivenScreenWidth()?

526avijitgupta replied:
Done.

AllanYangZhou commented on core/templates/dev/head/directives.js:
```diff
@@ -205,3 +205,26 @@ oppia.directive('mobileFriendlyTooltip', ['$timeout', function($timeout) {
+      element.on('click', function() {
+        if (!isDisabled) {
+          $('html, body').animate({
+            scrollTop: element.offset().top - 60
```
Recommend you first assign the `60` to a well named variable, so readers understand its purpose better.

526avijitgupta replied:
Took it out into a constant.

seanlip commented on core/templates/dev/head/dashboard/Dashboard.js:
```diff
@@ -50,25 +61,55 @@ oppia.controller('Dashboard', [
       $scope.checkForMobileView();
     });
 
+    $scope.toggleDropdown = function(type, explorationId, event) {
```
Hard to understand what this function does -- can you add docs about what 'type' represents?

526avijitgupta replied:
Done

seanlip commented on core/templates/dev/head/dashboard/Dashboard.js:
```diff
@@ -16,20 +16,31 @@
     };
 
+    $scope.isDropdownOpen = {};
+    $scope.statsDropdownOpened = {};
```
The name isn't enough to understand what this object is. Could you please add more documentation?

526avijitgupta replied:
Done

wxyxinyu commented on core/templates/dev/head/dashboard/Dashboard.js:
```diff
@@ -104,6 +153,7 @@ oppia.controller('Dashboard', [
       // This function is passed as a custom comparator function to `orderBy`,
       // so that special cases can be handled while sorting explorations.
       var value = entity[$scope.currentSortType];
+      console.log(value);
```
remove debug logging

526avijitgupta replied:
Had forgotten to push the commit later to this. Fixed.

wxyxinyu commented on core/templates/dev/head/dashboard/dashboard.html:
```diff
@@ -195,79 +195,150 @@ <h1 class="stat-value"><[dashboardStats.total_open_feedback]></h1>
       <div ng-if="myExplorationsView === 'card' && explorationsList.length > 0"
-           style="margin-top: 8px;">
-        <md-card ng-repeat="exploration in explorationsList track by exploration.id"
+           style="margin-top: 8px; position: relative;">
```
could these be put in css instead?

526avijitgupta replied:
Done

wxyxinyu commented:
One thing, could you truncate long feedback messages and unresolved answers?

wxyxinyu commented:
For mobile, could you hide the unhandled answers section, since it is not easily actionable by the creator? Since we will only have 2 sections, I think it will be better in the mobile case to just display the sections, without the dropdown thing, i.e.
<img width="295" alt="mock2" src="https://cloud.githubusercontent.com/assets/5061134/17848328/c3760b90-6807-11e6-9c77-7ae7d9cedca3.png">

wxyxinyu commented on core/templates/dev/head/dashboard/Dashboard.js:
```diff
@@ -37,24 +43,33 @@ oppia.constant('HUMAN_READABLE_EXPLORATIONS_SORT_BY_KEYS', {
+    $scope.activeExplorationId = '';
+    // Keeps track of the sub-dropdown that is opened in the main exploration
+    // dropdown.
+    $scope.statsDropdownOpened = {};
```
I have the same question about this -- This could, like `activeExplorationId` above, just be a string, couldn't it? At most one stats dropdown is opened.

Unless the intended behavior is that if I switch between the dropdowns of 2 explorations, which sub-dropdown is opened is preserved? I think this is not necessarily, however.

526avijitgupta replied:
Hmm. Done

wxyxinyu commented on core/templates/dev/head/dashboard/Dashboard.js:
```diff
@@ -73,25 +88,59 @@ oppia.controller('Dashboard', [
 
-    $scope.checkForMobileView();
+    $scope.activeExplorationIdOnMobile = function(explorationId) {
+      if ($scope.checkMobileView() &&
```
could just do 
`return $scope.checkMobileView() && $scope.activeExplorationId === explorationId;`

526avijitgupta replied:
Done

wxyxinyu commented:
I played with this a bit more, the interaction is very nice, especially the scrolling bit :)

526avijitgupta commented:
@wxyxinyu Truncated new feedback/unresolved answers section, with a word wrap. I've set the truncate value to 100 chars, since I believe we can afford displaying such lengthy values due to limiting the number of these displayed to just 3 for now!
Hid the unresolved answers section and made the other 2 sub dropdown (open feedback/last updates) forced open on mobile.

wxyxinyu commented:
![image](https://cloud.githubusercontent.com/assets/5061134/17882454/64f935d2-68c1-11e6-92a5-dcbef2fa70c4.png)
so from what I'm seeing, there's this gap between the feedback and updates, is it possible to remove that?
Also, the margin below the icons seem to be very large. (Right beside the dev mode label.) Could these be reduced?

AllanYangZhou commented:
Code changes looking good, once the layout issues Xinyu pointed out are resolved I can LGTM.

526avijitgupta commented:
@AllanYangZhou @wxyxinyu Sorry for the delay, will be working on this later in the day.

seanlip commented on index.yaml:
```diff
@@ -90,6 +97,35 @@ indexes:
 
 - kind: FeedbackMessageModel
   properties:
+  - name: <property object at 0x103ef5ba8>
```
This looks weird. What's going on here?

526avijitgupta replied:
Removed all changes to index.yaml

seanlip commented on core/storage/feedback/gae_models.py:
```diff
@@ -113,14 +113,30 @@ def get_by_exp_and_thread_id(cls, exploration_id, thread_id):
+        # http://stackoverflow.com/a/15552890
+        if len(exploration_ids) == 0:
+            return []
+        return cls.get_all().filter(
```
I'm worried about this. get_all() is going to be very expensive/inefficient in production, which is an issue since the creator dashboard is a commonly loaded page...

seanlip replied:
Sorry! I was mistaken here, get_all() is safe. The data is retrieved only on fetch().

That said, I was chatting with @wxyxinyu and we identified other issues in how the storage/queries are laid out -- they don't scale well as written. I think she'll be in touch about this to recommend next steps.

526avijitgupta commented:
@wxyxinyu @AllanYangZhou Fixed the layout issues as pointed out. Also, I was finally able to fix the error in backend tests!

seanlip commented:
OK -- not sure if @wxyxinyu had time to write, so I'll paraphrase what I remember and maybe she can fill in when she gets back from travelling.

Basically, the current logic looks like it gets all commits/feedback messages for all of a particular user's explorations, orders them all by last-updated, and shows the top N. It seems, however, that this is going to be quite labor-intensive, especially since the dashboard is loaded so frequently. In addition, what if the last 20 commits came from a single exploration? According to @wxyxinyu the intent was actually to show the last K (= 3?) commits/feedback per exploration.

So I think we discussed a couple of possibilities:
- show number of open feedback (as we did before) but only grab the actual feedback snippet data when we need it, e.g. if the user opens an exploration tile to see more
- maybe drop commit data for now until we have a more scalable way to get it

but I think the current approach is going to lead to significant issues in production due to it having to make a huge number of datastore gets when the dashboard page is loaded, so I'm a bit worried about it being merged.

To recap, this is a brief overview of the brief discussion I had with @wxyxinyu yesterday; she can probably fill in more detail. But perhaps you have some interim ideas/thoughts?

seanlip commented:
Oh, btw, what exactly was the issue with the backend tests? It's great that you fixed it, though I don't quite understand why the fix solved the problem!

526avijitgupta commented:
@seanlip My thoughts:
* Load feedback data only when the user clicks on it once: Something like separate route on the backend ? But we'll have to implement caching along with this, to prevent the huge amount of un-necessary requests that could be sent if the user toggles the dropdown multiple times!
* That sounds good. I'll open an issue in the issue tracker for this and reference this PR as well so that future contributors can look at the basic implementation if they like to!

About the backend tests, I used `git bisect` to find the commit where things broke. And then looking at the code in that commit made me realise I was accessing non-existent keys out of the last updates dict. AFAIR, it throws a 'KeyError' in such cases. Though I have no idea why this particular type of error was not present in the output logs of the failing tests!

seanlip commented:
Yes, I agree with a separate route on the backend + caching in the frontend. That makes sense.

And yep, what you say in the second bullet sounds good!

Interesting about the KeyError -- thanks for explaining! I'm surprised too that the error itself wasn't present. You don't happen to have a copy of the old test run by any chance, do you? If not, no worries, but if so, I thought I might have another look and figure out if it's something wrong with our testing system or if there's another explanation for it. But don't worry about it if you didn't save a copy!

seanlip commented:
Hi @526avijitgupta -- FYI, we are doing a release on Sept 3. Could we aim to do the per-exploration feedback thing before then, so that we can ship the new dashboard?

Thanks!

/cc @wxyxinyu @AllanYangZhou

526avijitgupta commented:
@seanlip Yes, I'll be able to complete the per-exploration feedback thing by then.
Also, about your previous comment, I was getting the same errors as on Travis, so they can be used as a reference!

526avijitgupta commented:
@seanlip @wxyxinyu @AllanYangZhou Please review.

wxyxinyu commented:
The code/approach looks good to me, but I have 1 small thing and 1 more major thing
Small thing: When you click on a tile and get the dropdown, it says "0 open feedback" at first, and then changes to the number of open feedback
More major thing: For anonymous feedback, it's labeled as "(by )". It should say "(by anonymous)" or something.

seanlip commented:
Maybe "(by a learner)"?

526avijitgupta commented:
@wxyxinyu Replies:
> Small thing: When you click on a tile and get the dropdown, it says "0 open feedback" at first, and then changes to the number of open feedback
Added a fallback text in the latest commit: "Retrieving [Answers/Feedback] Data".

> major thing: For anonymous feedback, it's labeled as "(by )". It should say "(by anonymous)" or something
Followed @seanlip 's suggestion.

seanlip commented on core/controllers/dashboard.py:
```diff
@@ -261,6 +266,20 @@ def get(self):
+        self.render_json({
+            'new_feedback': [
+                feedback_message.to_dict()
+                for feedback_message in feedback_services.get_messages_for_exp_id( # pylint: disable=line-too-long
```
You could actually break this line after "in".

526avijitgupta replied:
Done

seanlip commented on core/controllers/dashboard.py:
```diff
@@ -261,6 +266,20 @@ def get(self):
 
 
+class ExplorationStats(base.BaseHandler):
+    """Returns the new feedback and unresolved answers for an exploration."""
```
Seems to return only the feedback, not the answers?

526avijitgupta replied:
Done

seanlip commented on core/controllers/dashboard.py:
```diff
@@ -261,6 +266,20 @@ def get(self):
         })
 
 
+class ExplorationStats(base.BaseHandler):
```
Why is this called ExplorationStats? Maybe ExplorationDashboardDataHandler or something like that.

526avijitgupta replied:
Done

seanlip commented on core/domain/feedback_services.py:
```diff
@@ -237,6 +237,33 @@ def _get_thread_from_model(thread_model):
         thread_model.created_on, thread_model.last_updated)
 
 
+def get_messages_multi_for_exps(exploration_ids, limit=None):
```
Do we still need the 'multi' functionality? If not, I recommend removing it, and just restricting to retrieving messages for a single exp id. It's conceptually simpler, and I think this is the general workflow that we want to encourage anyway.

526avijitgupta replied:
Removed

seanlip commented on core/domain/stats_services.py:
```diff
@@ -40,50 +40,85 @@ def get_top_unresolved_answers_for_default_rule(exploration_id, state_name):
+            'count': 10 (number of unresolved answers for this exploration),
+            'unresolved_answers': (list of unresolved answers sorted by count)
+              [
+                {'count': 4, 'value': 'answer_1', 'state': 'Introduction'},
```
Some questions I have when reading this docstring:
- Are these counts expected to add up to the overall count?
- Is the full list of unresolved answers represented here, or is there truncation to e.g. the top 3?

526avijitgupta replied:
The list is truncated to top 3, so the counts are not expected to add up to the overall count as well!

seanlip replied:
Ah. It might be worth making that clear in the docstring.

526avijitgupta replied:
^Scratch that. Sorry, the individual counts should add up to the total count and the whole list is represented here. Edited the docstring to make this clear!

seanlip commented on core/domain/stats_services.py:
```diff
@@ -40,50 +40,85 @@ def get_top_unresolved_answers_for_default_rule(exploration_id, state_name):
+          ('eid2', 'Introduction'),
+          ('eid3', 'Introduction')
+        ]
+        when exp_ids = ['eid1', 'eid2', 'eid3']."""
```
Move """ to next line.

526avijitgupta replied:
Done

seanlip commented on core/domain/stats_services.py:
```diff
@@ -40,50 +40,85 @@ def get_top_unresolved_answers_for_default_rule(exploration_id, state_name):
+        explorations = (
+            exp_services.get_multiple_explorations_by_id(exp_ids, strict=False))
+        return [
+            (exp_domain_object.id, state_key)
```
exp_domain_object --> exploration
state_key --> state_name

to be consistent with how we describe these elsewhere in the codebase

526avijitgupta replied:
Done

seanlip commented on core/domain/stats_services.py:
```diff
@@ -40,50 +40,85 @@ def get_top_unresolved_answers_for_default_rule(exploration_id, state_name):
+            for state_key in exp_domain_object.states
+        ]
+
+    explorations_states_tuples = _get_explorations_state_tuples_by_ids(exp_ids)
```
To be consistent, the method should be _get_explorations_state**s**_tuples_by_ids

526avijitgupta replied:
Done

seanlip commented on core/storage/feedback/gae_models.py:
```diff
@@ -113,14 +113,30 @@ def get_by_exp_and_thread_id(cls, exploration_id, thread_id):
     @classmethod
     def get_threads(cls, exploration_id):
-        """Returns an array of threads associated to the exploration.
+        """Returns threads ordered by last updated associated to the
```
Returns a list of threads associated to the exploration, ordered by their last-updated field.

526avijitgupta replied:
Done

seanlip commented on core/storage/feedback/gae_models.py:
```diff
@@ -113,14 +113,30 @@ def get_by_exp_and_thread_id(cls, exploration_id, thread_id):
 
+    @classmethod
+    def get_threads_multi(cls, exploration_ids):
+        """Returns an array of threads ordered by last updated associated to
```
Returns an array of threads that ... hang on, do we actually still need this method?

526avijitgupta replied:
Done

526avijitgupta replied:
Removed.

seanlip commented:
Hi @526avijitgupta, @wxyxinyu asked me to do a second review of the backend. I _think_ the approach is fine, performance-wise, although I have a number of comments mostly regarding to getting rid of the methods/functions that fetch feedback threads for multiple exploration ids at a time.

@wxyxinyu, we should earmark this for production testing prior to the next release, just to see whether the dashboard performs at an acceptable speed for users who have a number of explorations.

AllanYangZhou commented on core/templates/dev/head/domain/dashboard/DashboardBackendApiService.js:
```diff
@@ -33,7 +33,26 @@ oppia.factory('DashboardBackendApiService', [
+      $http.get(requestUrl, {
+        cache: true
+      }).then(function(response) {
+        if (resolve) {
```
I don't think the `if` on `resolve` and `reject` are necessary. The `$q` library should always pass in these args, no?

AllanYangZhou replied:
On further inspection, I think constructing a promise in the first place is entirely unnecessary here. `http.get` already returns a promise, and promises are composable, so you shouldn't need to wrap around it with another promise.

AllanYangZhou replied:
To be clearer, you should be able to do this:

    http.get(url).then(function(response) {
      return response.data;
    }).then(function(data) {
      console.log(data) // equal to response.data
    })

526avijitgupta replied:
Done. Also, reading a few parts of the code at https://github.com/angular/angular.js/blob/master/src/ng/http.js lead me to believe that angular's `$http` already uses `$q` internally!

526avijitgupta replied:
@AllanYangZhou Please see latest code. I've implemented the simplest implementation - return the `$http.get()` promise directly from the service, and then use it inside controller as required!

526avijitgupta commented:
@seanlip @AllanYangZhou @wxyxinyu Please review.

seanlip commented on core/controllers/dashboard.py:
```diff
@@ -262,6 +267,21 @@ def get(self):
+            'new_feedback': [
+                feedback_message.to_dict()
+                for feedback_message in
+                feedback_services.get_messages_for_exp_id(
```
Do you need to return all? Or just top 3?

526avijitgupta replied:
Had missed passing the `limit` parameter. Fixed now.

seanlip commented on core/controllers/dashboard.py:
```diff
@@ -262,6 +267,21 @@ def get(self):
+    """Returns the new feedback for an exploration."""
+
+    @base.require_fully_signed_up
+    def get(self):
```
Why not include exploration_id in the URL? That's standard with most of our handlers.

526avijitgupta replied:
Sorry, I don't think I get you here. The exploration ID is in the URL already!

seanlip replied:
Use something like '/.../exploration_id>'

and then you can change this method signature to get(self, exploration_id)

526avijitgupta replied:
Done

seanlip commented on core/storage/feedback/gae_models.py:
```diff
@@ -183,13 +184,16 @@ def get(cls, exploration_id, thread_id, message_id, strict=True):
 
     @classmethod
-    def get_messages(cls, exploration_id, thread_id):
+    def get_messages(cls, exploration_id, thread_id, is_full_thread_id=False):
```
Oh, goodness, no, please don't do this. We are going to end up with a whole lot of confusion about what's the full thread id and what's the thread id.

Keep the API as is, and do any conversions needed before calling it.

526avijitgupta replied:
Oops, my bad. Looks like I chose the easy way out, which turned out to be unsuitable for this case.
Fixed this now - removed the flag and added a property to the model to get the thread id.

seanlip commented on core/storage/feedback/gae_models.py:
```diff
@@ -183,13 +184,16 @@ def get(cls, exploration_id, thread_id, message_id, strict=True):
+            FeedbackThreadModel.generate_full_thread_id(
+                exploration_id, thread_id)
+            if not is_full_thread_id else thread_id)
+        print full_thread_id
```
Please remove print statements.

526avijitgupta replied:
Done

seanlip commented on core/domain/stats_services.py:
```diff
@@ -40,50 +40,88 @@ def get_top_unresolved_answers_for_default_rule(exploration_id, state_name):
-    Note that this method currently returns the counts only for the DEFAULT
-    rule. This should ideally handle all types of unresolved answers.
+    Note that this method currently returns the data only for the DEFAULT rule.
+    This should ideally handle all types of unresolved answers.
```
Make this sentence into a TODO(526avijitgupta).

526avijitgupta replied:
Done

seanlip commented on core/domain/feedback_services.py:
```diff
@@ -366,6 +366,34 @@ def _get_thread_from_model(thread_model):
+                exp_id, thread.id, is_full_thread_id=True))
+
+    result = []
+    for message in message_models:
```
message --> message_model

526avijitgupta replied:
Done. Also, we would've preferred to write "message" directly if it were a domain object, right ?

seanlip replied:
Yes, that's right.

seanlip commented on core/domain/feedback_services.py:
```diff
@@ -366,6 +366,34 @@ def _get_thread_from_model(thread_model):
+    thread_models = feedback_models.FeedbackThreadModel.get_threads(exp_id)
+
+    message_models = []
+    for thread in thread_models:
```
thread --> thread_model

526avijitgupta replied:
Done

seanlip commented on core/domain/feedback_services.py:
```diff
@@ -366,6 +366,34 @@ def _get_thread_from_model(thread_model):
         thread_model.created_on, thread_model.last_updated)
 
 
+def get_messages_for_exp_id(exp_id, limit=None):
```
The way this is implemented makes it look like it won't scale well to explorations that have many threads/messages. Is it possible to do something like: query the messages datastore for stuff with exp_id matching the given exp_id, and sort backwards by last_updated, and limit to top 3 or 10 or something? That's going to be more efficient than fetching and iterating through every single message each time.

526avijitgupta replied:
`exploration_id` isn't a property of a FeedbackMessage object. Such an object is represented by it's thread_id. Thus, I don't think directly querying the messages using exploration_id is possible.

As a simple fix, we could limit the no. of messages fetched through the `get_messages` query to `feconf.NEW_FEEDBACK_COUNT_DASHBOARD` instead of the default query limit:

```diff
@@ -198,7 +198,7 @@ class FeedbackMessageModel(base_models.BaseModel):
                 exploration_id, thread_id))

         return cls.get_all().filter(
-            cls.thread_id == full_thread_id).fetch(feconf.DEFAULT_QUERY_LIMIT)
+            cls.thread_id == full_thread_id).fetch(feconf.NEW_FEEDBACK_COUNT_DASHBOARD)
```

seanlip replied:
This is probably OK -- it doesn't fully fix the issue, but you at least limit to 3 messages (say) per thread. That said, would it make more sense to show at most one message per open thread -- just the last one? If so you could just get the N most recently-updated threads and then grab their last messages.

526avijitgupta replied:
If the dashboard mentions "New Feedback", then as a user I would expect that the feedback messages be latest (irrespective of the thread they belong to). Thus, I think that might not be a very good option.

seanlip replied:
But is it likely that multiple feedback messages from the same thread will give the user more info about what's going on than just one message? Normally one thread would be about one issue (like on github).

526avijitgupta replied:
Thinking about this again - I agree that increasing the datastore queries is too big a compromise for this. Thus I've followed your original suggestion - removed the limit parameter from the feedback_services method, created a new method for the model's API and used it to fetch the latest 3 feedback threads and finally extracted the most recent message out of them and sent this to the frontend - so it's now a lot less no. of queries to the datastore!

seanlip commented on core/domain/stats_services.py:
```diff
@@ -40,50 +40,88 @@ def get_top_unresolved_answers_for_default_rule(exploration_id, state_name):
     return exps_answers_mapping
 
 
+def get_exp_unresolved_answers_for_default_rule(exp_id):
```
Where is this used?

526avijitgupta replied:
Removed.

seanlip commented on feconf.py:
```diff
@@ -505,6 +506,11 @@ def get_empty_ratings():
 SHOW_TRAINABLE_UNRESOLVED_ANSWERS = False
+# Number of unresolved answers to be displayed in the dashboard for each
+# exploration.
+TOP_UNRESOLVED_ANSWERS_COUNT_DASHBOARD = 3
```
If we only display 3, why do we have functions that seem to fetch all?

526avijitgupta replied:
All are fetched by only a single `get_multi` query, and then limited to 3 before sending in the payload to frontend: https://github.com/oppia/oppia/blob/6bae5592f79810e04474a9fe501fb0b9fc707db4/core/domain/stats_services.py#L89-L90

seanlip replied:
Is it possible at all to limit earlier? Unnecessary datastore fetches cost time/money, so it seems worth optimizing since the dashboard page is quite frequently used.

526avijitgupta replied:
It should be ideally possible. But looking at the data pipeline - it'll need some more investigation.
`stats_services.get_top_state_rule_answers_multi` calls `stats_domain.StateRuleAnswerLog.get_multi_by_multi_explorations` internally which in turn calls `stats_models.StateRuleAnswerLogModel.get_or_create_multi_for_multi_explorations` where the actual fetch from the datastore seems to happen. Looking at the code for `stats_models.StateRuleAnswerLogModel.get_or_create_multi_for_multi_explorations`: https://github.com/oppia/oppia/blob/2aeb715c5595c1676c8a56fe67449959c01dca7d/core/storage/statistics/gae_models.py#L94-L145 .. it isn't immediately clear to me how to limit to the top N results by frequency!

seanlip replied:
A possibility is to do some pre-computing using e.g. continuous computations. I think that's what ultimately would need to happen but we'll need to be a bit careful to remove handled answers from that list, so maybe better to push to a future PR.

In the meantime, though, what do you think about perhaps showing the unresolved answers in the same way as we do the feedback threads -- i.e. only when a particular exploration is examined? Then we can file an issue to say: "show top answers without having to click on each exploration" or something, and that's the issue which requires pre-computation to be set up.

Also, just as a sanity check to inform this decision: let's say a dashboard has 50 explorations each with 10 states, and each state has maybe 1000 answers (where there are 200 unique answers). Can you estimate roughly how many datastore fetches are made, and how large the data transferred would be?

526avijitgupta replied:
I get your point about loading the unresolved answers on click. Considering your own example - will any user go ahead and click on each of his/her 50 explorations to see their stats ? No. In most cases he/she might click on a few or even not click on any of them at all! (After all, it's important to consider that every type of user, and not just creators are redirected to the dashboard when they open https://oppia.org, assuming they are logged in). So it makes sense to load the bare minimum data for rendering the dashboard page and keeping it as fast as possible!
^I do believe that it would've been better if this realization would've come to me earlier 😛 

Also, that's a whole lot of queries in the above example (200 * 10 * 50 = 100000 unique unresolved answers!) which leads to an even bigger amount of data being transferred.

526avijitgupta replied:
@seanlip  I just realized that this implementation was already a part of the Version 2.3.2 released on 8th August, so it might be even live! Looking at the part of the code for calculating the count of unresolved answers per exploration - the same method is being used: https://github.com/oppia/oppia/pull/2101/files#diff-eb73e4e06cbd9d6da9674f8310fb8004R70 . Thus I guess we might have to fix this earlier than we previously intended!

seanlip replied:
Hm, so, just to check: can you confirm whether or not the changes to the unresolved answers stuff in this PR would result in a latency that's no worse than what's already on develop?

526avijitgupta replied:
I can confirm that the latency wouldn't be worse than what's already in develop. And that's because the no. of datastore queries remains the same, it's just that we need a bit of extra computation of data this time!

seanlip replied:
OK! In that case, I'm fine with deferring a fix for this to a separate PR. Let's just make sure an issue is filed for it.

seanlip commented on feconf.py:
```diff
@@ -505,6 +506,11 @@ def get_empty_ratings():
+# exploration.
+TOP_UNRESOLVED_ANSWERS_COUNT_DASHBOARD = 3
+# Number of open feedback to be displayed in the dashboard for each exploration.
+NEW_FEEDBACK_COUNT_DASHBOARD = 3
```
Where is this used?

526avijitgupta replied:
Ah, had missed this. Passed as value to limit param

AllanYangZhou commented on core/templates/dev/head/dashboard/Dashboard.js:
```diff
@@ -73,6 +88,26 @@ oppia.controller('Dashboard', [
+          function(response) {
+            $scope.explorationStats[explorationId] = response.data;
+          }, function(errorResponse) {
+            if (FATAL_ERROR_CODES.indexOf(errorResponse.status) !== -1) {
```
What happens if this isn't true? Perhaps we should add another warning in that case, instead of swallowing the error.

526avijitgupta replied:
@AllanYangZhou This seems be to a subset of https://github.com/oppia/oppia/issues/2142 . To address this issue here, I've added a constant to base.js and used that as a warning everywhere when the request fails without these error codes! (See https://github.com/oppia/oppia/pull/2296/commits/c4b7e7233e10b07d797a51fc04fa2ee0170bac20)

526avijitgupta replied:
@AllanYangZhou Just realized that there's a minor issue with this - how do I detect if the request failed because the server has taken too long to respond ? Currently I'm (wrongly) assuming that if the request's status codes is not in the list of fatal error codes then it might be due to the timeout but I think this is a very vague way of implementing this!
A quick google search lead me to http://stackoverflow.com/questions/17873961/determine-request-time-out-in-angular-http - but I'm not sure if any of the answers here address this situation well enough!
/cc @seanlip

seanlip replied:
I think, don't handle timeout for now. Let's handle it globally, later, as part of #2142. The right thing to do isn't obvious.

526avijitgupta replied:
Done

AllanYangZhou commented on core/templates/dev/head/dashboard/Dashboard.js:
```diff
@@ -121,32 +177,29 @@ oppia.controller('Dashboard', [
-      function(errorStatus) {
-        if (FATAL_ERROR_CODES.indexOf(errorStatus) !== -1) {
+      function(errorResponse) {
+        if (FATAL_ERROR_CODES.indexOf(errorResponse.status) !== -1) {
```
Same question as above for the error codes.

526avijitgupta replied:
Same as above.

AllanYangZhou commented:
One UI related question that just occurred to me: is there a reason we don't simply bring the user to the feedback page when they click on open feedback? Specifically, I am thinking when people click here:
![ex_mouse](https://cloud.githubusercontent.com/assets/8732147/18755361/cff26182-80a0-11e6-944b-6dedfb9a3486.png)

seanlip commented on scripts/install_third_party.sh:
```diff
@@ -122,27 +122,27 @@ if [ ! -d "$TOOLS_DIR/webtest-1.4.2" ]; then
-echo Checking if browsermob-proxy is installed in $TOOLS_DIR/pip_packages
-if [ ! -d "$TOOLS_DIR/browsermob-proxy-0.7.1" ]; then
-  echo Installing browsermob-proxy
+# # Python API for browsermob-proxy.
```
Don't commit this.

526avijitgupta replied:
Sorry, reverted.

526avijitgupta commented:
(Changed the title of the PR so that it better reflects on what has been done as well as https://github.com/oppia/oppia/issues/1366 is closed when this gets in!)

526avijitgupta commented:
@seanlip I'll open up an issue for the unresolved answers part. Apart from that, I've addressed all the review comments.
 @AllanYangZhou That's a good idea. I've implemented clicking on the feedback's author text would take to the feedback page.

Anything else left to be done before this gets in ?

seanlip commented on core/domain/feedback_services.py:
```diff
@@ -366,6 +366,30 @@ def _get_thread_from_model(thread_model):
 
+def get_new_messages_for_exp_id(exp_id):
+    """Fetch the feedback threads for the given exploration, and use those to
+    finally get the new (open) feedback messages.
```
I traced this code path, but I don't see where "open" comes into it. Where are you filtering on 'open'?

Also, I wonder if we could standardize the naming a bit. We have get_new_messages_for_exp_id() which calls get_latest_updated_threads(). Maybe they should be get_most_recent_messages() and get_most_recently_updated_threads()?

526avijitgupta replied:
Actually we use "open" and "new" interchangeably wrt to the dashboard. That's because what we display as New Feedback on the dashboard frontend is actually the value of open feedback. 
Standardizing the naming seems like a good idea, I'll do that.

seanlip replied:
Ah, I see. For that sort of thing, then, the standard way to do it is to use 'open' everywhere except the final point in the UI, at which you change it to 'new'. That way you don't have multiple words in the backend code that mean the same thing.

526avijitgupta replied:
Done

seanlip commented on core/domain/feedback_services.py:
```diff
@@ -366,6 +366,30 @@ def _get_thread_from_model(thread_model):
+       A list of FeedbackMessage.
+    """
+    thread_models = (
+        feedback_models.FeedbackThreadModel.get_latest_updated_threads(exp_id))
```
Should this be something like get_open_threads()?

526avijitgupta replied:
I think `get_most_recently_updated_threads()` would be a better name, as per your suggestion in the comment above.

526avijitgupta replied:
Renamed to `get_most_recently_updated_threads()`.

seanlip commented on core/domain/feedback_services.py:
```diff
@@ -366,6 +366,30 @@ def _get_thread_from_model(thread_model):
+    message_models = []
+    for thread_model in thread_models:
+        message_models.append(
+            feedback_models.FeedbackMessageModel.get_most_recent_message(
```
Would a get_multi version of this result in any savings?

526avijitgupta replied:
I don't think so - since a get_multi version would actually be passed in a list of thread ids for which we want the most recent message out of, and it'll fetch the most recent messages for which `cls.thread_id.IN(thread_ids_list)` followed by `.order(-cls.last_updated)`. This resultant list would be mostly larger in size (since it'll not be ensured that there is just a single message out of each thread) and thus could might even take more time than 3 simple get calls!
or am I missing something ?

seanlip replied:
Nope, the reasoning sounds fine. Just wanted to check.

seanlip commented on core/storage/feedback/gae_models.py:
```diff
@@ -113,14 +117,26 @@ def get_by_exp_and_thread_id(cls, exploration_id, thread_id):
                 cls.last_updated).fetch(feconf.DEFAULT_QUERY_LIMIT)
 
+    @classmethod
+    def get_latest_updated_threads(cls, exploration_id):
```
This method and the previous one look almost identical. Why not pass in a 'limit' arg to the previous one which defaults to feconf.DEFAULT_QUERY_LIMIT?

526avijitgupta replied:
Done.

seanlip commented on index.yaml:
```diff
@@ -57,6 +57,13 @@ indexes:
 
 - kind: ExplorationCommitLogEntryModel
   properties:
+  - name: deleted
```
I don't see any additional calls to ExplorationCommitLogEntryModel in this PR. Why is the additional index entry needed?

526avijitgupta replied:
This might've gotten updated in an earlier stage of this PR. Reverted now.

seanlip commented:
Oh, btw, I only did a backend review. @AllanYangZhou can you PTAL at the frontend?

Thanks!

seanlip commented:
Backend looks good to me! Thanks @526avijitgupta .

@AllanYangZhou or @wxyxinyu, could you PTAL at frontend?

wxyxinyu commented on core/templates/dev/head/pages/dashboard/dashboard.html:
```diff
@@ -194,79 +196,133 @@ <h1 class="stat-value"><[dashboardStats.total_open_feedback]></h1>
+            <div class="stat-block">
+              <div class="heading-text"
+                   ng-click="toggleSubDropdown(EXPLORATION_DROPDOWN_STATS.OPEN_FEEDBACK, $event)">
+                <[explorationStats[exploration.id] ? explorationStats[exploration.id].open_feedback.length + ' Open Feedback': 'Retrieving Feedback Data ...']>
```
It seems like this is causing the dashboard to display "3 Open Feedback" even when there are more than 3. If there are more than 3, you should probably use the feedback count used at the top of the dashboard page, and also indicate that the feedback is truncated (perhaps with "and x more" at the bottom).

526avijitgupta replied:
Great catch! Done.

wxyxinyu commented on core/templates/dev/head/pages/dashboard/dashboard.html:
```diff
@@ -194,79 +196,133 @@ <h1 class="stat-value"><[dashboardStats.total_open_feedback]></h1>
+            <div class="stat-block" ng-if="!checkMobileView()">
+              <div class="heading-text"
+                   ng-click="toggleSubDropdown(EXPLORATION_DROPDOWN_STATS.TOP_UNRESOLVED_ANSWERS, $event)">
+                <[explorationStats[exploration.id] ? exploration.top_unresolved_answers.length + ' Unresolved Answers from users': 'Retrieving Answers Data ...']>
```
Similar to above for unresolved answers, you should use the total number of unresolved answers, and not the length of the data.

526avijitgupta replied:
Done

wxyxinyu commented on core/templates/dev/head/pages/dashboard/dashboard.html:
```diff
@@ -60,8 +60,10 @@ <h1 class="stat-value"><[dashboardStats.total_open_feedback]></h1>
           </a>
         </li>
-        {% if can_create_collections %}
+        {% if not can_create_collections %}
```
Hmm, what's the purpose of adding this check?

526avijitgupta replied:
This has been added as a layout fix when My Collections tab isn't visible. Else we'll have to use `float: right` to keep things in position!

wxyxinyu replied:
I see, thanks!

seanlip replied:
Can we add a comment in the file to explain this? It's not obvious.

526avijitgupta replied:
Done

526avijitgupta commented:
@wxyxinyu Resolved the comments.
@seanlip I think you too will have to approve this PR, since it currently shows 2 reviews requesting changes.

seanlip commented:
Approved from my end (for backend only)

526avijitgupta commented:
@wxyxinyu Fixed the incorrect value of count. I guess this should be ready to go in now!

wxyxinyu commented on core/templates/dev/head/directives.js:
```diff
@@ -162,3 +162,29 @@ oppia.directive('mobileFriendlyTooltip', ['$timeout', function($timeout) {
   };
 }]);
+
+oppia.directive('scrollOnClick', function() {
```
Just clarified with @BenHenning -- we should not be adding new directives to directives.js. Instead, you should put it into a new file in core/templates/dev/head/domain/dashboard/, for example core/templates/dev/head/domain/dashboard/ScrollOnClickDirective.js

This is because this scroll directive is very specific to the dashboard dropdown.

Sorry for not picking up on this earlier!

526avijitgupta replied:
Done

526avijitgupta commented:
@wxyxinyu Please review.
